File: backend/src/lib/db/models/progress.ts
import { db } from '../client';
import { UserTopic, UserLessonProgress } from '../../types/database';

export class ProgressModel {
  static async createUserTopic(
    userId: string,
    topicId: number,
    goalText: string,
    targetDate?: Date
  ): Promise<UserTopic> {
    const timestamp = new Date();
    const result = await db.execute({
      sql: `INSERT INTO user_topics (user_id, topic_id, goal_text, target_date, started_at, last_accessed)
            VALUES (?, ?, ?, ?, ?, ?)
            RETURNING *`,
      args: [
        userId,
        topicId,
        goalText,
        targetDate ? targetDate.toISOString() : null,
        timestamp.toISOString(),
        timestamp.toISOString()
      ]
    });

    return this.mapUserTopic(result.rows[0]);
  }

  static async getUserTopic(userId: string, topicId: number): Promise<UserTopic | null> {
    const result = await db.execute({
      sql: 'SELECT * FROM user_topics WHERE user_id = ? AND topic_id = ?',
      args: [userId, topicId]
    });

    return result.rows[0] ? this.mapUserTopic(result.rows[0]) : null;
  }

  static async updateUserTopic(
    userId: string,
    topicId: number,
    updates: Partial<UserTopic>
  ): Promise<UserTopic | null> {
    const fields: string[] = [];
    const args: any[] = [];

    if (updates.goalText) {
      fields.push('goal_text = ?');
      args.push(updates.goalText);
    }

    if (updates.targetDate) {
      fields.push('target_date = ?');
      args.push(updates.targetDate.toISOString());
    }

    if (updates.currentLessonId) {
      fields.push('current_lesson_id = ?');
      args.push(updates.currentLessonId);
    }

    if (fields.length === 0) {
      return null;
    }

    fields.push('last_accessed = ?');
    args.push(new Date().toISOString());

    args.push(userId);
    args.push(topicId);

    const result = await db.execute({
      sql: `UPDATE user_topics 
            SET ${fields.join(', ')}
            WHERE user_id = ? AND topic_id = ?
            RETURNING *`,
      args
    });

    return result.rows[0] ? this.mapUserTopic(result.rows[0]) : null;
  }

  static async createLessonProgress(
    userId: string,
    lessonId: number,
    status: string
  ): Promise<UserLessonProgress> {
    const timestamp = Date.now();
    const result = await db.execute({
      sql: `INSERT INTO lesson_progress (user_id, lesson_id, status, time_spent_minutes, last_interaction_at, created_at, updated_at)
            VALUES (?, ?, ?, ?, ?, ?, ?)
            RETURNING *`,
      args: [userId, lessonId, status, 0, timestamp, timestamp, timestamp]
    });

    return this.mapLessonProgress(result.rows[0]);
  }

  static async getLessonProgress(
    userId: string,
    lessonId: number
  ): Promise<UserLessonProgress | null> {
    const result = await db.execute({
      sql: 'SELECT * FROM lesson_progress WHERE user_id = ? AND lesson_id = ?',
      args: [userId, lessonId]
    });

    return result.rows[0] ? this.mapLessonProgress(result.rows[0]) : null;
  }

  static async updateLessonProgress(
    userId: string,
    lessonId: number,
    status: string
  ): Promise<UserLessonProgress> {
    const timestamp = Date.now();
    const result = await db.execute({
      sql: `UPDATE lesson_progress
            SET status = ?, last_interaction_at = ?, updated_at = ?
            WHERE user_id = ? AND lesson_id = ?
            RETURNING *`,
      args: [status, timestamp, timestamp, userId, lessonId]
    });

    return this.mapLessonProgress(result.rows[0]);
  }

  private static mapUserTopic(row: any): UserTopic {
    return {
      userId: row.user_id,
      topicId: row.topic_id,
      currentLessonId: row.current_lesson_id,
      goalText: row.goal_text,
      targetDate: row.target_date ? new Date(row.target_date) : null,
      startedAt: new Date(row.started_at),
      lastAccessed: new Date(row.last_accessed)
    };
  }

  private static mapLessonProgress(row: any): UserLessonProgress {
    return {
      userId: row.user_id,
      lessonId: row.lesson_id,
      status: row.status,
      timeSpentMinutes: row.time_spent_minutes || 0,
      lastInteractionAt: row.last_interaction_at,
      createdAt: row.created_at,
      updatedAt: row.updated_at
    };
  }
}

File: backend/src/lib/db/models/progress.py
from datetime import datetime
from typing import List, Optional
from backend.src.lib.db.client import dbClient
from backend.src.lib.utils.auth import generateId
from backend.src.lib.utils.dates import now, to_unix_timestamp, from_unix_timestamp
from shared.src.types import Progress, TopicProgress

class ProgressModel:
    @staticmethod
    async def create(
        userId: str,
        topicId: str,
        questionId: str,
        isCorrect: bool
    ) -> Progress:
        id = generateId()
        timestamp = to_unix_timestamp(now())

        try:
            result = await dbClient.execute(
                sql="""
                    INSERT INTO user_progress (id, user_id, topic_id, question_id, is_correct, created_at)
                    VALUES (?, ?, ?, ?, ?, ?)
                    RETURNING *
                """,
                args=[id, userId, topicId, questionId, 1 if isCorrect else 0, timestamp]
            )

            if not result.rows or len(result.rows) == 0:
                raise Exception('Failed to create progress entry')

            return ProgressModel.mapProgress(result.rows[0])
        except Exception as e:
            print('Error creating progress:', e)
            raise e

    @staticmethod
    async def getByTopicId(userId: str, topicId: str) -> List[Progress]:
        try:
            result = await dbClient.execute(
                sql='SELECT * FROM user_progress WHERE user_id = ? AND topic_id = ? ORDER BY created_at DESC',
                args=[userId, topicId]
            )

            return [ProgressModel.mapProgress(row) for row in (result.rows or [])]
        except Exception as e:
            print('Error getting progress:', e)
            raise e

    @staticmethod
    async def getTopicProgress(topicId: str) -> TopicProgress:
        try:
            result = await dbClient.execute(
                sql="""
                    SELECT
                        (SELECT COUNT(*) FROM user_progress WHERE topic_id = ? AND is_correct = 1) as correct_answers,
                        (SELECT COUNT(*) FROM user_progress WHERE topic_id = ? AND is_correct = 0) as incorrect_answers,
                        (SELECT COUNT(*) FROM questions WHERE topic_id = ?) as total_questions,
                        (SELECT ROUND((julianday('now') - julianday(MIN(created_at))) * 24 * 60)
                         FROM user_progress
                         WHERE topic_id = ?) as time_spent_minutes
                """,
                args=[topicId, topicId, topicId, topicId]
            )

            row = result.rows[0] if result.rows else None
            return {
                'correctAnswers': int(row.correct_answers if row and row.correct_answers else 0),
                'incorrectAnswers': int(row.incorrect_answers if row and row.incorrect_answers else 0),
                'totalQuestions': int(row.total_questions if row and row.total_questions else 0),
                'timeSpentMinutes': int(row.time_spent_minutes if row and row.time_spent_minutes else 0)
            }
        except Exception as e:
            print('Error getting topic progress:', e)
            raise e

    @staticmethod
    def mapProgress(row: any) -> Progress:
        try:
            return {
                'id': row.id,
                'userId': row.user_id,
                'topicId': row.topic_id,
                'questionId': row.question_id,
                'isCorrect': bool(row.is_correct),
                'createdAt': from_unix_timestamp(row.created_at)
            }
        except Exception as e:
            print('Error mapping progress:', e)
            raise e
